SqlInstall.ps1

```diff
@@ -0,0 +1,433 @@
+
+    .PARAMETER Authentication will prompt you if you want mixed mode authentication or just Windows AD authentication. With Mixed Mode, you will be prompted for the SA password.
+
+    .PARAMETER EnableException
+            By default, when something goes wrong we try to catch it, interpret it and give you a friendly warning message.
+            This avoids overwhelming you with "sea of red" exceptions, but is inconvenient because it basically disables advanced scripting.
+            Using this switch turns this "nice by default" feature off and enables you to catch exceptions with your own try/catch.
+
+    .Inputs
+    None
+
+    .Outputs
+    None
+
+    .Example
+    C:\PS> Install-DbaInstance
+
+    This will run the installation with the default settings
+
+    .Example
+
+    C:\PS> Install-DbaInstance -AppVolume "G"
+
+    This will run the installation with default setting apart from the application volume, this will be redirected to the G drive.
+
+    .Example 
+
+    C:\PS> Install-DbaInstance -Version 2016 -AppVolume "D" -DataVolume "E" -LogVolume "L" -PerformVolumeMaintenance "Yes" -SqlServerAccount "MyDomain\SvcSqlServer"
+
+    This will install SQL Server 2016 on the D drive, the data on E, the logs on L and the other files on the autodetected drives. The perform volume maintenance
+    right is granted and the domain account SvcSqlServer will be used as the service account for SqlServer.
+
+
+    #>
+    Param  (
+        [ValidateSet("2008", "2008R2", "2012", "2014", "2016", "2017", "2019")][string]$Version, 
+        [ValidateSet("Express", "Standard", "Enterprise", "Developer")][string]$Edition,
+        [ValidateSet("AllFeaturesWithDefaults", "Custom")][string]$Role,
+        [ValidateSet("Python", "R", "Python and R")][string]$StatsAndMl,
+        [string]$AppVolume, 
+        [string]$DataVolume, 
+        [string]$LogVolume, 
+        [string]$TempVolume, 
+        [string]$BackupVolume,
+        [string]$InstallFolder,
+        [ValidateSet("Yes", "No")][string]$PerformVolumeMaintenance,
+        [ValidateSet("Yes")][string]$SaveFile,
+        [ValidateSet("Windows", "Mixed Mode")][string]$Authentication,
+        [switch]$EnableException
+    )
+
+
+    if ($Null -eq $Version -or $Version -eq '') {
+        Stop-Function -Message "You need to specify a SQL Server Version to run this function." -Continue -EnableException $EnableException
+    }
+
+    If($Null -eq $Edition -or $Edition -eq '')
+    {
+        $Edition = "Express"
+    }
+
+    If($null -eq $Role -or $Role -eq "AllFeaturesWithDefaults"){
+        If($null -eq $Edition)
+        {
+            $Edition = "Express"
+        }
+    }
+
+    # Check if the edition of SQL Server supports Python and R. Introduced in SQL 2016, it should not be allowed in earlier installations.
+
+    IF ( $null -eq $StatsAndMl -or $StatsAndMl -ne '' ) {
+        $Array = "2016", "2017", "2019"
+
+        IF ($Version -notin $Array ) {
+            $StatsAndMl = '';
```

Comment:
What if they pass StatsAndMl and 2008? This should be a check with a `Stop-Function`

Also, you'll love `-not`, it's super convenient. Takes care of nulls and empties.

`if (-not $StatsAndMl -or -not $StatsAndMl) {`

Also make sure you use lower case for `begin`, `process`, `end`, `if`, `else`, `foreach` and so on.

Reply:
There are more combinations that shouldn't work. Even with more parameters to come...